web/datapipeline/sap_systems_projector.go

```diff
@@ -0,0 +1,84 @@
+package datapipeline
+
+import (
+	"fmt"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/trento-project/trento/internal/sapsystem"
+	"github.com/trento-project/trento/web/entities"
+	"gorm.io/gorm"
+	"gorm.io/gorm/clause"
+)
+
+func NewSAPSystemsProjector(db *gorm.DB) *projector {
+	SAPSystemsProjector := NewProjector("sapsystems", db)
+
+	SAPSystemsProjector.AddHandler(SAPsystemDiscovery, SAPSystemsProjector_SAPSystemsDiscoveryHandler)
+
+	return SAPSystemsProjector
+}
+
+func SAPSystemsProjector_SAPSystemsDiscoveryHandler(dataCollectedEvent *DataCollectedEvent, db *gorm.DB) error {
+	decoder := getPayloadDecoder(dataCollectedEvent.Payload)
+	var discoveredSAPSystems sapsystem.SAPSystemsList
+	if err := decoder.Decode(&discoveredSAPSystems); err != nil {
+		log.Errorf("can't decode data: %s", err)
+		return err
+	}
+
+	for _, s := range discoveredSAPSystems {
+		instance := entities.SAPSystemInstance{
+			AgentID:  dataCollectedEvent.AgentID,
+			SystemID: s.Id,
+			SID:      s.SID,
+		}
+
+		if s.Type == 1 {
+			instance.Type = sapsystem.SAPSystemsDatabase
+		}
+
+		if s.Type == 2 {
+			instance.Type = sapsystem.SAPSystemsApplication
+			instance.DBHost = fmt.Sprint(s.Profile["SAPDBHOST"])
+		}
+
+		for _, i := range s.Instances {
+			var features string
+			var instanceNumber string
+
+			if p, ok := i.SAPControl.Properties["SAPSYSTEM"]; ok {
+				instanceNumber = p.Value
+			} else {
+				continue
+			}
+
+			for _, SAPControlInstance := range i.SAPControl.Instances {
+				if instanceNumber == fmt.Sprintf("%02d", SAPControlInstance.InstanceNr) {
+					features = SAPControlInstance.Features
```

Comment:
We can exit from the loop once this condition is met, as we won't never have it as true again for this sap control instance, or instance in general. Let's add a break

Reply:
I don't like breaks, I do this only because I love you